File: packages/react-native-video/ios/hybrids/VideoViewViewManager/HybridVideoViewViewManager.swift
//
//  HybridVideoViewViewManager.swift
//  ReactNativeVideo
//
//  Created by Krzysztof Moch on 23/09/2024.
//

import Foundation
import AVKit
import NitroModules

struct ViewListenerPair {
  let id: UUID
  let eventName: String
  let callback: Any
}

class HybridVideoViewViewManager: HybridVideoViewViewManagerSpec {
  weak var view: VideoComponentView?
  var listeners: [ViewListenerPair] = []
  
  let DEALOCATED_WARNING = "ReactNativeVideo: VideoComponentView is no longer available. It is likely that the view was deallocated."
  
  init(nitroId: Double) throws {
    guard let view = VideoComponentView.globalViewsMap.object(forKey: NSNumber(value: nitroId)) else {
      throw VideoViewError.viewNotFound(nitroId: nitroId).error()
    }
    
    self.view = view
    super.init()
    view.delegate = VideoViewDelegate(viewManager: self)
  }
  
  // MARK: - Private helpers
  
  private func addListener<T>(eventName: String, listener: T) -> ListenerSubscription {
    let id = UUID()
    listeners.append(ViewListenerPair(id: id, eventName: eventName, callback: listener))
    return ListenerSubscription(remove: { [weak self] in
      self?.listeners.removeAll { $0.id == id }
    })
  }
  
  private func emitEvent<T>(eventName: String, invoke: (T) throws -> Void) {
    for pair in listeners where pair.eventName == eventName {
      if let callback = pair.callback as? T {
        do {
          try invoke(callback)
        } catch {
          print("[ReactNativeVideo] Error calling \(eventName) listener: \(error)")
        }
      } else {
        print("[ReactNativeVideo] Invalid callback type for \(eventName)")
      }
    }
  }
  
  // MARK: - Properties
  
  weak var player: (any HybridVideoPlayerSpec)? {
    get {
      guard let view = view else {
        print(DEALOCATED_WARNING)
        return nil
      }
      return view.player
    }
    set {
      guard let view = view else {
        print(DEALOCATED_WARNING)
        return
      }
      view.player = newValue
    }
  }
  
  var controls: Bool {
    get {
      guard let view else {
        print(DEALOCATED_WARNING)
        return false
      }
      
      return view.controls
    }
    set {
      guard let view else {
        print(DEALOCATED_WARNING)
        return
      }
      
      view.controls = newValue
    }
  }
  
  var pictureInPicture: Bool {
    get {
      guard let view else {
        print(DEALOCATED_WARNING)
        return false
      }
      
      return view.allowsPictureInPicturePlayback
    }
    set {
      guard let view else {
        print(DEALOCATED_WARNING)
        return
      }
      
      view.allowsPictureInPicturePlayback = newValue
    }
  }
  
  var autoEnterPictureInPicture: Bool {
    get {
      guard let view else {
        print(DEALOCATED_WARNING)
        return false
      }
      
      return view.autoEnterPictureInPicture
    }
    set {
      guard let view else {
        print(DEALOCATED_WARNING)
        return
      }
      
      view.autoEnterPictureInPicture = newValue
    }
  }
  
  var resizeMode: ResizeMode {
    get {
      guard let view else {
        print(DEALOCATED_WARNING)
        return .none
      }
      
      return view.resizeMode
    }
    set {
      guard let view else {
        print(DEALOCATED_WARNING)
        return
      }
      
      view.resizeMode = newValue
    }
  }
  
  var keepScreenAwake: Bool {
    get {
      guard let view else {
        print(DEALOCATED_WARNING)
        return false
      }
      
      return view.keepScreenAwake
    }
    set {
      guard let view else {
        print(DEALOCATED_WARNING)
        return
      }
      
      view.keepScreenAwake = newValue
    }
  }
  
  // Android only - no-op on iOS
  var surfaceType: SurfaceType = .surface
  
  func enterFullscreen() throws {
    guard let view else {
      throw VideoViewError.viewIsDeallocated.error()
    }
    
    try view.enterFullscreen()
  }
  
  func exitFullscreen() throws {
    guard let view else {
      throw VideoViewError.viewIsDeallocated.error()
    }
    
    try view.exitFullscreen()
  }
  
  func enterPictureInPicture() throws {
    guard let view else {
      throw VideoViewError.viewIsDeallocated.error()
    }
    
    try view.startPictureInPicture()
  }
  
  func exitPictureInPicture() throws {
    guard let view else {
      throw VideoViewError.viewIsDeallocated.error()
    }
    
    try view.stopPictureInPicture()
  }
  
  func canEnterPictureInPicture() -> Bool {
    return AVPictureInPictureController.isPictureInPictureSupported()
  }
  
  // MARK: - Listener registration methods
  
  func addOnPictureInPictureChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onPictureInPictureChange", listener: listener)
  }
  
  func addOnFullscreenChangeListener(listener: @escaping (Bool) -> Void) throws -> ListenerSubscription {
    addListener(eventName: "onFullscreenChange", listener: listener)
  }
  
  func addWillEnterFullscreenListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willEnterFullscreen", listener: listener)
  }
  
  func addWillExitFullscreenListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willExitFullscreen", listener: listener)
  }
  
  func addWillEnterPictureInPictureListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willEnterPictureInPicture", listener: listener)
  }
  
  func addWillExitPictureInPictureListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    addListener(eventName: "willExitPictureInPicture", listener: listener)
  }
  
  func clearAllListeners() throws {
    listeners.removeAll()
  }
  
  // MARK: - Event emission methods
  
  func onPictureInPictureChange(_ isActive: Bool) {
    emitEvent(eventName: "onPictureInPictureChange") { (callback: (Bool) throws -> Void) in try callback(isActive) }
  }
  
  func onFullscreenChange(_ isActive: Bool) {
    emitEvent(eventName: "onFullscreenChange") { (callback: (Bool) throws -> Void) in try callback(isActive) }
  }
  
  func willEnterFullscreen() {
    emitEvent(eventName: "willEnterFullscreen") { (callback: () throws -> Void) in try callback() }
  }
  
  func willExitFullscreen() {
    emitEvent(eventName: "willExitFullscreen") { (callback: () throws -> Void) in try callback() }
  }
  
  func willEnterPictureInPicture() {
    emitEvent(eventName: "willEnterPictureInPicture") { (callback: () throws -> Void) in try callback() }
  }
  
  func willExitPictureInPicture() {
    emitEvent(eventName: "willExitPictureInPicture") { (callback: () throws -> Void) in try callback() }
  }
}

File: packages/react-native-video/android/src/main/java/com/twg/video/hybrids/videoviewviewmanager/HybridVideoViewViewManager.kt
package com.margelo.nitro.video

import androidx.annotation.OptIn
import androidx.media3.common.util.UnstableApi
import com.facebook.proguard.annotations.DoNotStrip
import com.twg.video.core.VideoManager
import com.twg.video.core.VideoViewError
import com.twg.video.core.utils.PictureInPictureUtils
import com.twg.video.core.utils.Threading
import java.util.UUID

data class ViewListenerPair(
  val id: UUID,
  val eventName: String,
  val callback: Any
)

@DoNotStrip
@OptIn(UnstableApi::class)
class HybridVideoViewViewManager(nitroId: Int): HybridVideoViewViewManagerSpec(), VideoViewEventsEmitter {
  private var videoView =
    VideoManager.getVideoViewWeakReferenceByNitroId(nitroId) ?: throw VideoViewError.ViewNotFound(nitroId)
  private val listeners = mutableListOf<ViewListenerPair>()

  init {
    videoView.get()?.eventsEmitter = this
  }

  override var player: HybridVideoPlayerSpec?
    get() {
      return Threading.runOnMainThreadSync { return@runOnMainThreadSync videoView.get()?.hybridPlayer }
    }
    set(value) {
      Threading.runOnMainThread {
        videoView.get()?.hybridPlayer = value as? HybridVideoPlayer
      }
    }

  override fun canEnterPictureInPicture(): Boolean {
    return PictureInPictureUtils.canEnterPictureInPicture()
  }

  override fun enterFullscreen() {
    videoView.get()?.enterFullscreen()
  }

  override fun exitFullscreen() {
    videoView.get()?.exitFullscreen()
  }

  override fun enterPictureInPicture() {
    Threading.runOnMainThread {
      videoView.get()?.enterPictureInPicture()
    }
  }

  override fun exitPictureInPicture() {
    Threading.runOnMainThread {
      videoView.get()?.exitPictureInPicture()
    }
  }

  override var autoEnterPictureInPicture: Boolean
    get() = videoView.get()?.autoEnterPictureInPicture == true
    set(value) {
      Threading.runOnMainThread {
        videoView.get()?.autoEnterPictureInPicture = value
      }
    }

  override var pictureInPicture: Boolean
    get() = videoView.get()?.pictureInPictureEnabled == true
    set(value) {
      videoView.get()?.pictureInPictureEnabled = value
    }

  override var controls: Boolean
    get() = videoView.get()?.useController == true
    set(value) {
      videoView.get()?.useController = value
    }

  override var resizeMode: ResizeMode
    get() = videoView.get()?.resizeMode ?: ResizeMode.NONE
    set(value) {
      videoView.get()?.resizeMode = value
    }

  override var keepScreenAwake: Boolean
    get() = videoView.get()?.keepScreenAwake == true
    set(value) {
      videoView.get()?.keepScreenAwake = value
    }

  override var surfaceType: SurfaceType
    get() = videoView.get()?.surfaceType ?: SurfaceType.SURFACE
    set(value) {
      videoView.get()?.surfaceType = value
    }

  // MARK: - Private helpers

  private fun <T> addListener(eventName: String, listener: T): ListenerSubscription {
    val id = UUID.randomUUID()
    listeners.add(ViewListenerPair(id, eventName, listener as Any))
    return ListenerSubscription { listeners.removeAll { it.id == id } }
  }

  @Suppress("UNCHECKED_CAST")
  private fun <T> emitEvent(eventName: String, invoke: (T) -> Unit) {
    listeners.filter { it.eventName == eventName }.forEach { pair ->
      try {
        invoke(pair.callback as T)
      } catch (e: Exception) {
        println("[ReactNativeVideo] Error calling $eventName listener: $e")
      }
    }
  }

  // MARK: - Listener registration methods

  override fun addOnPictureInPictureChangeListener(listener: (Boolean) -> Unit): ListenerSubscription {
    return addListener("onPictureInPictureChange", listener)
  }

  override fun addOnFullscreenChangeListener(listener: (Boolean) -> Unit): ListenerSubscription {
    return addListener("onFullscreenChange", listener)
  }

  override fun addWillEnterFullscreenListener(listener: () -> Unit): ListenerSubscription {
    return addListener("willEnterFullscreen", listener)
  }

  override fun addWillExitFullscreenListener(listener: () -> Unit): ListenerSubscription {
    return addListener("willExitFullscreen", listener)
  }

  override fun addWillEnterPictureInPictureListener(listener: () -> Unit): ListenerSubscription {
    return addListener("willEnterPictureInPicture", listener)
  }

  override fun addWillExitPictureInPictureListener(listener: () -> Unit): ListenerSubscription {
    return addListener("willExitPictureInPicture", listener)
  }

  override fun clearAllListeners() {
    listeners.clear()
  }

  // MARK: - Event emission methods (called by VideoView)

  override fun onPictureInPictureChange(isActive: Boolean) {
    emitEvent<(Boolean) -> Unit>("onPictureInPictureChange") { it(isActive) }
  }

  override fun onFullscreenChange(isActive: Boolean) {
    emitEvent<(Boolean) -> Unit>("onFullscreenChange") { it(isActive) }
  }

  override fun willEnterFullscreen() {
    emitEvent<() -> Unit>("willEnterFullscreen") { it() }
  }

  override fun willExitFullscreen() {
    emitEvent<() -> Unit>("willExitFullscreen") { it() }
  }

  override fun willEnterPictureInPicture() {
    emitEvent<() -> Unit>("willEnterPictureInPicture") { it() }
  }

  override fun willExitPictureInPicture() {
    emitEvent<() -> Unit>("willExitPictureInPicture") { it() }
  }

  override val memorySize: Long
    get() = 0
}

interface VideoViewEventsEmitter {
  fun onPictureInPictureChange(isActive: Boolean)
  fun onFullscreenChange(isActive: Boolean)
  fun willEnterFullscreen()
  fun willExitFullscreen()
  fun willEnterPictureInPicture()
  fun willExitPictureInPicture()
}
